ow_system_plugins/base/bol/seo_service.php

```diff
@@ -130,7 +144,160 @@ public function getSitemapPath()
      */
     public function getBaseSitemapPath()
     {
-        return OW::getPluginManager()->getPlugin('base')->getUserFilesDir() . self::SITEMAP_DIR_NAME . '/';
+        $path = OW::getPluginManager()->getPlugin('base')->getUserFilesDir() . self::SITEMAP_DIR_NAME . '/';
+
+        if ( !file_exists($path) )
+        {
+            mkdir($path, 0777);
+        }
+
+        return $path;
+    }
+
+    /**
+     * Build sitemap
+     *
+     * @return void
+     */
+    public function buildSitemap()
+    {
+        $urls = $this->sitemapDao->findUrlList(self::SITEMAP_MAX_URLS_IN_FILE);
+        $sitemapBuild = (int) OW::getConfig()->getValue('base', 'seo_sitemap_last_build') + 1;
+        $entities = $this->getSitemapEntities();
+        $sitemapIndex = (int) OW::getConfig()->getValue('base', 'seo_sitemap_index');
+        $sitemapPath = $this->getBaseSitemapPath() . $sitemapBuild . '/';
+
+        if ( !file_exists($sitemapPath) )
+        {
+            mkdir($sitemapPath, 0777);
+        }
+
+        if ( $urls )
+        {
+            // generate parts of sitemap
+            $processedUrls = [];
+            $allLanguages    = BOL_LanguageService::getInstance()->getLanguages();
+            $currentLanguage = null;
+
+            // get active languages
+            $activeLanguages = array();
+            foreach($allLanguages as $language)
+            {
+                if ( $language->status == 'active' )
+                {
+                    $activeLanguages[] = $language;
+
+                    if ( $language->order == 1 )
+                    {
+                        $currentLanguage = $language->id;
+                    }
+                }
+            }
+
+            $activeLanguagesCount = count($activeLanguages);
+
+            // process urls
+            foreach( $urls as $urlDto )
```

Comment:
Replace all invalid chars in URL list for sitemap. Info is here - http://www.sitemaps.org/protocol.html